PROTOSPEC.md

```diff
@@ -0,0 +1,62 @@
+
+There is a special outer scope for built-in decorators. When a decorator is declared at the module or script level, it is being declared in an inner, nested scope which may shadow things in this outer scope. In this way, built-in decorators can be polyfilled.
+
+## Early errors
+
+Although the decorators specification is designed to allow decorators to be analyzed ahead of time, the JavaScript specification is organized with everything running at runtime, except for analysis of early errors. Because decorators may be imported from other modules, decorators are specified to operate entirely runtime, rather than at early error time.
+
+## Runtime representation of decorators
+
+A decorator as applied is represented as a List of Decorator Records of the form { [[BuiltinDecorator]]: the String name of a built-in decorator, e.g., `"wrap"`, [[Arguments]]: a List of JavaScript values }.
+
+A `decorator` declaration is, effectively, a function which converts a List of JavaScript values into a List of Decorator Records.
+
+Note, the length of the list and the [[BuiltinDecorator]] fields can be pre-computed at parse time, if the imports are parsed. Only the [[Arguments]] will differ across multiple runs of the program.
+
+## Execution
+
+When evaluating decorators, there are two stages:
+1. First, when evaluating the class declaration, the DecoratorLists are evaluated into a list 
+2. Then, later in the class evaluation, the the built-in decorators which resulted from the previousl evaluation are invoked directly, at various points in time.
+
+### Evaluating a decorator list
+
+Decorator lists are evaluated interspersed with evaluation of the class (e.g., with respect to computed property names). In practice, at runtime, what's really being evaluated is the argument lists. The evaluation basically works like ordinary function calls, bottoming out when a built-in decorator is reached.
+1. Let *result* be an empty list.
+1. For each *decorator* in the DecoratorList:
+  1. Evaluate the arguments of *decorator*.
+  1. If *decorator* is a built-in decorator, append it as a Decorator Record to *result*.
+  1. Otherwise, find the `decorator @thisDecorator` definition and call it, which will give a new DecoratorList. Recurse at Step 2.
+
+### Applying built-in decorators
+
+Decorator lists are evaluated into lists of built-in decorators. These are accumulated on the class in ways that are specific to each decorator:
+- For `@register`, a class-wide list of calls to `register` is constructed (with the list including both the function to call and, if applicable, the property descriptor to pass to it). This list is called, in order (inside to outside, top to bottom), as the last step of creating a class.
+- For `@initialize`, the field definition Record (as output by ClassFieldDefinitionEvaluation) is annotated with the callback passed, and then the callback is run each time the initializer is run. Note, only one `@initialize` call may be made per field.
```

Comment:
Why multiple registers but only one initialize?

Reply:
Multiple `@register`s has important use cases, e.g., unrelated pieces of metadata, but `@initialize` sort of logically takes over the field declaration. We could make it work, though; you could file an issue about this feature request if you can explain why you want that. I could imagine consistency arguments in either direction.

Reply:
Consistency and "principle of least surprise" would be mine; since i might not have any idea that the decorators i'm using is using `initialize`, and i explicitly slap one on the bottom, i'd expect it to override the initialization.

Reply:
You could imagine multiple semantics here, either removing the other `@initialize` call, or using its return value, etc. Sometimes we conservatively ban things when intuition can lead us astray. This is what I meant when i said I could imagine consistency arguments in either direction. It's hard to think about this abstractly, without concrete use cases in mind; feel free to file an issue on this question.

Reply:
@littledan, what about multiple decorators that require instance-level initializers? 
E.g., in my project, I have a system where my decorators can exchange property keys between each other. It looks like the following:
```javascript
// propertyDecorator.js

import {apiMap} from './utils';

const api = (descriptor) => {
  return {
    ...descriptor,
    extras: [
      {
        initializer() {
          apiMap.set(this, desriptor.key);
        }
        kind: 'initializer',
        placement: 'static',
      },
    ],
  }
}

// classDecorator.js
import {apiMap} from './utils';

const cls = ({elements, key}) => {
  let $api;

  return {
    elements: [
      ...elements,
      {
        finisher() {
          $api = apiMap.get(this);
        },
        kind: 'initializer',
        placement: 'static',
      }
    ],
    key,
  }
}
```
User can apply a lot of different decorators for the single field, and many of them use this approach. The restriction "one `@initialize` per field" could break this system and reduce the number of opportunities previous proposal provided. 

As I understand, the `@initialize` is the unique approach to pierce into a `constructor` of the class that inherits idea of the `{ kind: 'initializer' }` of the previous proposal. If I am correct, it should have more power than to serve as a simple field initializer. E.g., we could use it for `@bound` decorator.

Reply:
Thanks for this example. Given that the finishers are static, I'm wondering, would the `@register` decorator work for you? If not, do you think you could give a little more context here, so I can understand the example?

Reply:
@littledan, oh I see, all of what I'm talking about is proposed in [NEXTBUILTINS.md](https://github.com/tc39/proposal-decorators/blob/static/NEXTBUILTINS.md#bound).

Reply:
Well, which of those should be prioritized, from your perspective? `@register` is part of the proposed MVP.

Reply:
@littledan, GitHub didn't show me your answers for some reason 😁. I've created an issue with suggestion (#260) for the one thing that I consider important for my case.

Reply:
> If not, do you think you could give a little more context here, so I can understand the example?

Sorry, I'm not sure why I have chosen this example for the explanation. I thought about instance-level, but initializers in my example are completely declaration-level. I should've chosen another example.